### ISSUE TITLE
stty in your config.fish causes externally backgrounded fish instances to stop

### ISSUE THREAD

**Issue by jorgebucaran** [fish-shell/fish-shell#5246]
Using `stty` in your `config.fish` causes background processes to suddenly stop.

To reproduce the issue add to your `config.fish` the following code.

```
stty werase undef
```

Then send any job to the background using `&`.

```fish
$ fish -c "echo" &
Job 1, 'fish -c "echo" &' has stopped
```

Running `jobs` tells me:
```console
Job	Group	State	Command
1	83624	stopped	fish -c "echo" &
```

I can bring the process to the foreground using `fg` to finish it.

```fish
$ fg
Send job 1, 'fish -c "echo" &' to foreground
stty: tcsetattr: Interrupted system call
```

I know that wrapping the `stty` command like so:

```fish
if status --is-interactive
    ...
end
```
...will prevent the issue, but it's unfriendly to tell people how to write their configuration.

This issue was discovered by @Tyilo while using jorgebucaran/fisher (see https://github.com/jorgebucaran/fisher/issues/469). The presence of `stty` in their configuration causes `fisher` to hang because `fisher` keeps waiting for a process that is stopped.

I've been able to reproduce this issue on macOS, Linux (Arch) / xterm-256color running fish 2.0.0, 2.2.0, 2.7.1, and the latest fish from HEAD.

**Comment by mqudsi** [comment]
This is unfortunate since explicitly using stty in config.fish is the current workaround to #2980. 

As an aside, the title made me understand that it breaks background jobs within the session, not externally backgrounded fish runs.

**Comment by jorgebucaran** [comment]
Thanks for having a look @mqudsi. I updated the title.

**Comment by mqudsi** [comment]
It's not `stty`, it's anything that needs to read/write from/to the terminal. Adding `cat` to your `config.fish` and then trying to run it in the background does the same.

I'm not sure this is a bug? It's a backgrounded process (i.e. a process that does NOT have control of the terminal) trying to access the terminal, it's going to `SIGTTOU` or `SIGTTIN` until it's brought to the foreground and given control of the shell. What do you expect to happen?

**Comment by mqudsi** [comment]
Additionally, this is really independent of `config.fish`, even. Try running `fish -c 'cat' &` in a terminal, and you'll see the same behavior:

```
mqudsi@Blitzkrieg /m/d/G/iMessage> fish -c 'cat' &
Job 1, 'fish -c 'cat' &' has stopped
```

or even bash:

```
mqudsi@Blitzkrieg /m/d/G/iMessage> bash -c 'cat' &
Job 1, 'bash -c 'cat' &' has stopped
```

**Comment by jorgebucaran** [comment]
@mqudsi The backgrounded process is not created in `config.fish`.

**Comment by mqudsi** [comment]
I don't understand, can you explain what you mean? stty is being run as part of the background fish instance, isn't it?

**Comment by jorgebucaran** [comment]
> This issue was discovered by @Tyilo with Fisher (see jorgebucaran/fisher#469). The presence of stty in their configuration causes fisher to hang because fisher keeps waiting for a process that is stopped.

Fisher is not doing anything with `stty`. Fisher forks `fish` in the background for each package is trying to download, enabling concurrent downloads. The problem is the user is/was using `stty` in their `config.fish`. I don't know the reason why they are using `stty`. The problem is that using `stty` in `config.fish` will cause backgrounded fish instances to stop / hang.

**Comment by jorgebucaran** [comment]
> This issue was discovered by @Tyilo with Fisher (see jorgebucaran/fisher#469). The presence of stty in their configuration causes fisher to hang because fisher keeps waiting for a process that is stopped.

Fisher is not doing anything with `stty`. Fisher forks `fish` in the background for each package is trying to download, enabling concurrent downloads. The problem is the user is/was using `stty` in their `config.fish`. I don't know the reason why they are using `stty`. The problem is that using `stty` in `config.fish` will cause backgrounded fish instances to stop / hang.

cc @mqudsi

**Comment by jorgebucaran** [comment]
> @mqudsi It's not stty, it's anything that needs to read/write from/to the terminal.

Thanks for the clarification. Is there anyway to solve this? Should I tell users (albeit less casually) "Please do not put `stty` in your `config.fish` or something along those lines?

**Comment by mqudsi** [comment]
@jorgebucaran I just want to make sure you're clear that it's not strictly an "`stty` in `config.fish`" issue. _Any_ process that attempts to modify or read from the terminal is going to run into the same issue, although perhaps `stty` is special because people don't realize it's doing if they're not clear on how it works internally.

@faho @zanchey @ridiculousfish just a thought here, but what if we started shipping `config.fish` with something like the following:

```fish
# Place general fish and environment configuration here:

# Anything beyond this point is not executed when fish isn't run interactively, such as via `fish -c '...'`
if not status --is-interactive
    exit
end

# Place any external commands or commands modifying the behavior of terminal IO here
```

**Comment by jorgebucaran** [comment]
@mqudsi Thanks for the clarification.

**Comment by mqudsi** [comment]
@jorgebucaran why don’t you fork fish without a home directory or with a known configuration? That way you don’t need to work about the user doing anything untoward.

**Comment by jorgebucaran** [comment]
@mqudsi Maybe that will create new unforeseen problems, but how can I do that anyway?

**Comment by jorgebucaran** [comment]
> @mqudsi ...why don’t you fork fish without a home directory or with a known configuration?

How? 👀 

Maybe that'll create new problems, but it's worth looking into.

**Comment by faho** [comment]
>@faho @zanchey @ridiculousfish just a thought here, but what if we started shipping config.fish with something like the following:

I've considered not reading configuration if we're not interactive, but that would probably break the fisher code here (if it uses a function defined in a snippet).

Even if we added that bit, I don't think users will get that `stty` touches the terminal, or that touching the terminal is bad if you're in the background. I don't think this particular user even knew what that was - this specific invocation did precisely nothing in fish.

So that warning about "terminal IO" won't really help much.

**Comment by jorgebucaran** [comment]
> @faho ...that would probably break the fisher code here (if it uses a function defined in a snippet).

Could you elaborate a bit more? I'm don't see why. 🤔 

> I don't think this particular user even knew what that was - this specific invocation did precisely nothing in fish.

Actually, in all this time, I never took the moment to ask @Tyilo: why where you using `stty` in your config.fish? Just curious.

Perhaps he had a valid use case?

**Comment by faho** [comment]
>Could you elaborate a bit more? I'm don't see why. thinking

If we don't source the configuration - neither config.fish, nor conf.d snippets - then you can't use the functions and variables defined there. So you'd be restricted in the kind of thing you could call in `fish -c`.

>Perhaps he had a valid use case?

Because that call is something that's useful in bash. My guess is they just copied it over.

**Comment by jorgebucaran** [comment]
> @faho ... nor conf.d snippets

Interesting. I didn't know this would mean not sourcing configuration snippets.

**Comment by mqudsi** [comment]
Yes, my idea was to explicitly skip the configuration file then source a "fisher jail configuration" that would allow you to set up an environment that would do what you need and load any functions, etc. necessary to proceed with the operation. It's more work to get started but it'd be entirely deterministic.

**Comment by jorgebucaran** [comment]
> @mqudsi why don’t you fork fish without a home directory or with a known configuration? 

How?

**Comment by mqudsi** [comment]
There's (currently) no command-line option to skip the user's `config.fish` (although it might be a good idea), but depending on what you execute you may be able to get away with redefining `HOME` before launching fish.

```fish
env HOME=/dev/null fish /path/to/fisher_download_script.fish &
```

will execute without a user-modified configuration and without any "known" configuration. To instead use a custom fisher environment, you can point `HOME` to a pre-installed directory containing whatever functions your fisher scripts need in a `.config/fish/functions` subdirectory of that directory.

Make sure to restore `HOME` for any external commands you run within your script, though!

**Comment by jorgebucaran** [comment]
Thanks, @mqudsi. I'll definitely try it out.

**Comment by jorgebucaran** [comment]
@mqudsi or @faho Could you summarize the position of the fish-shell developers regarding this issue (for future reference)? Poorly informed users like @Tyilo may conclude that what is happening here is a problem in Fisher's end, when in fact it isn't.

**Comment by faho** [comment]
>Could you summarize the position of the fish-shell developers regarding this issue (for future reference)?

My "position" is that currently, you need to guard all commands that "touch the terminal" (that means they produce output or try to fiddle with the terminal settings) with `status is-interactive`.

We're open to changes here. A few possibilities:

- Don't read configuration if not interactive (might be too restrictive?)

- Add a commandline-option to start fish without configuration (means you need to use it)

- Don't allow configuration access to the terminal if not interactive (half-baked)

But until then, not guarding is a user error.
